Colour is wrong in Intel QSV H265-10bit encodes

### Problem Description

When encoding H265-10bit video with Intel QSV, the resulting video has wrong colours -- seems to lack green.
This does not happen with QSV H265-8bit or with x265-10bit.

As I mentioned on the Reddit thread, the one notable difference between 8-bit and 10-bit encode logs is the line:
`Format (format=p010le)`

A sample of the source can be found [here](https://drive.google.com/file/d/1NdTyNeJDAtjyACNyWJvEPy1m9ofzCUXa/view?usp=share_link).
A sample encode [here](https://drive.google.com/file/d/1I_osuqp39nqLPnizs9qGs90MQ9RnemE4/view?usp=share_link).
I'm attaching sample screenshots (top is source, bottom encode).
![Source](https://user-images.githubusercontent.com/21968492/225067128-7e206491-6aac-4e2e-a359-1a88cec0bbac.png)
![Encode](https://user-images.githubusercontent.com/21968492/225067145-24d76f7c-8d59-46fd-bbf2-ea98c2c6a96a.png)

Note that this is a B/W video, but I've confirmed it with a full-colour Bluray -- the same lack of green (just less obvious among all the colours).

(Please ask github to add Matroska to the list of allowed formats...)

### Activity Log, Crash Log or any other details

```text
HandBrake 1.6.1 (2023012300)
OS: Microsoft Windows NT 10.0.19045.0
CPU: Intel(R) Pentium(R) CPU G4560 @ 3.50GHz
Ram: 8064 MB, 
GPU Information:
  Intel(R) HD Graphics 610 - 31.0.101.2111
Screen: 1920x1080
Temp Dir: C:\Users\Rosemary\AppData\Local\Temp\
Install Dir: C:\Program Files\HandBrake\
Data Dir: C:\Users\Rosemary\AppData\Roaming\HandBrake

-------------------------------------------

 # Starting Encode ...

[17:37:31] base preset: ESPer (Modified)
[17:37:31] hb_init: starting libhb thread
[17:37:31] Starting work at: Tue Mar 14 17:37:31 2023
[17:37:31] 1 job(s) to process
[17:37:31] json job:
{
  "Audio": {
    "AudioList": [
      {
        "Bitrate": -1,
        "DRC": 0,
        "Encoder": "flac16",
        "Gain": 0,
        "Mixdown": 4,
        "NormalizeMixLevel": false,
        "Samplerate": 0,
        "Track": 0,
        "DitherMethod": 0
      }
    ],
    "CopyMask": [
      "copy:aac",
      "copy:ac3"
    ]
  },
  "Destination": {
    "ChapterList": [
      {
        "Name": "Chapter 1"
      }
    ],
    "ChapterMarkers": true,
    "AlignAVStart": false,
    "File": "C:\\Users\\Rosemary\\Downloads\\Encode-QSV-H265-10bit.mkv",
    "Mp4Options": {
      "IpodAtom": false,
      "Mp4Optimize": false
    },
    "Mux": "av_mkv"
  },
  "Filters": {
    "FilterList": [
      {
        "ID": 4,
        "Settings": {
          "mode": "7"
        }
      },
      {
        "ID": 3,
        "Settings": {
          "block-height": "16",
          "block-thresh": "40",
          "block-width": "16",
          "filter-mode": "2",
          "mode": "3",
          "motion-thresh": "1",
          "spatial-metric": "2",
          "spatial-thresh": "1"
        }
      },
      {
        "ID": 14,
        "Settings": {
          "crop-bottom": "22",
          "crop-left": "0",
          "crop-right": "0",
          "crop-top": "22",
          "height": "1036",
          "width": "1920"
        }
      },
      {
        "ID": 7,
        "Settings": {
          "mode": "0"
        }
      }
    ]
  },
  "PAR": {
    "Num": 1,
    "Den": 1
  },
  "Metadata": {},
  "SequenceID": 0,
  "Source": {
    "Angle": 1,
    "Range": {
      "Type": "chapter",
      "Start": 1,
      "End": 1
    },
    "Title": 1,
    "Path": "C:\\Users\\Rosemary\\Downloads\\Source-10s.mkv"
  },
  "Subtitle": {
    "Search": {
      "Burn": false,
      "Default": false,
      "Enable": false,
      "Forced": false
    },
    "SubtitleList": []
  },
  "Video": {
    "Encoder": "qsv_h265_10bit",
    "Level": "auto",
    "Bitrate": 10000,
    "TwoPass": false,
    "Turbo": false,
    "ColorMatrixCode": 0,
    "Options": "",
    "Preset": "quality",
    "Profile": "auto",
    "QSV": {
      "Decode": true
    },
    "HardwareDecode": 0
  }
}
[17:37:31] CPU: Intel(R) Pentium(R) CPU G4560 @ 3.50GHz
[17:37:31]  - Intel microarchitecture Kaby Lake
[17:37:31]  - logical processor count: 4
[17:37:31] Intel Quick Sync Video support: yes, gpu list: 0
[17:37:31] Intel Quick Sync Video integrated adapter with index 0
[17:37:31] Impl mfxhw64 library path: C:\WINDOWS\System32\DriverStore\FileRepository\iigd_dch.inf_amd64_51f685305808e3a5\libmfxhw64.dll
[17:37:31]  - Intel Media SDK hardware: API 1.35 (minimum: 1.3)
[17:37:31]  - Decode support:  h264 hevc (8bit: yes, 10bit: yes)
[17:37:31]  - H.264 encoder: yes
[17:37:31]     - preferred implementation: hardware (1) via D3D11
[17:37:31]     - capabilities (hardware):  breftype icq+la+i+downs vsinfo chromalocinfo opt1 opt2+mbbrc+extbrc+trellis+repeatpps+ib_adapt+nmpslice
[17:37:31]  - H.265 encoder: yes (8bit: yes, 10bit: yes)
[17:37:31]     - preferred implementation: hardware (1) via D3D11
[17:37:31]     - capabilities (hardware):  bpyramid icq vsinfo masteringinfo cllinfo opt1
[17:37:31]  - AV1 encoder: no
[17:37:31] hb_scan: path=C:\Users\Rosemary\Downloads\Source-10s.mkv, title_index=1
udfread ERROR: ECMA 167 Volume Recognition failed
src/libbluray/disc/disc.c:333: failed opening UDF image C:\Users\Rosemary\Downloads\Source-10s.mkv
src/libbluray/disc/disc.c:437: error opening file BDMV\index.bdmv
src/libbluray/disc/disc.c:437: error opening file BDMV\BACKUP\index.bdmv
[17:37:31] bd: not a bd - trying as a stream/file instead
libdvdread: DVDOpenFileUDF:UDFFindFile /VIDEO_TS/VIDEO_TS.IFO failed
libdvdnav: vm: vm: failed to read VIDEO_TS.IFO
[17:37:31] dvd: not a dvd - trying as a stream/file instead
Input #0, matroska,webm, from 'C:\Users\Rosemary\Downloads\Source-10s.mkv':
  Metadata:
    encoder         : libebml v1.4.4 + libmatroska v1.7.1
    creation_time   : 2023-03-14T15:34:48.000000Z
  Duration: 00:00:10.31, start: 0.000000, bitrate: 35869 kb/s
  Stream #0:0: Video: h264 (High), yuv420p(tv, bt709, progressive), 1920x1080 [SAR 1:1 DAR 16:9], 23.98 fps, 23.98 tbr, 1k tbn (default)
    Metadata:
      BPS             : 33439801
      DURATION        : 00:00:10.303000000
      NUMBER_OF_FRAMES: 247
      NUMBER_OF_BYTES : 43066284
      _STATISTICS_WRITING_APP: mkvmerge v73.0.0 ('25 or 6 to 4') 64-bit
      _STATISTICS_WRITING_DATE_UTC: 2023-03-14 15:34:48
      _STATISTICS_TAGS: BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES
  Stream #0:1(jpn): Audio: dts (DTS-HD MA), 48000 Hz, 5.1(side), s32p (24 bit) (default)
    Metadata:
      BPS             : 2431270
      DURATION        : 00:00:10.304000000
      NUMBER_OF_FRAMES: 966
      NUMBER_OF_BYTES : 3131476
      _STATISTICS_WRITING_APP: mkvmerge v73.0.0 ('25 or 6 to 4') 64-bit
      _STATISTICS_WRITING_DATE_UTC: 2023-03-14 15:34:48
      _STATISTICS_TAGS: BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES
  Stream #0:2(eng): Subtitle: hdmv_pgs_subtitle (default)
    Metadata:
      BPS             : 0
      DURATION        : 00:00:00.000000000
      NUMBER_OF_FRAMES: 0
      NUMBER_OF_BYTES : 0
      _STATISTICS_WRITING_APP: mkvmerge v73.0.0 ('25 or 6 to 4') 64-bit
      _STATISTICS_WRITING_DATE_UTC: 2023-03-14 15:34:48
      _STATISTICS_TAGS: BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES
[17:37:31] scan: decoding previews for title 1
[17:37:31] scan: audio 0x1: dca, rate=48000Hz, bitrate=1 ??? (DTS-HD MA) (5.1 ch)
[17:37:32] scan: 10 previews, 1920x1080, 23.976 fps, autocrop = 22/22/0/0, aspect 16:9, PAR 1:1, color profile: 1-1-1, chroma location: left
[17:37:32] scan: supported video decoders: avcodec qsv
[17:37:32] libhb: scan thread found 1 valid title(s)
[17:37:32] qsv: integrated qsv adapter with index 0 has been selected
[17:37:32] Starting Task: Encoding Pass
[17:37:32] Skipping vfr filter
[17:37:32] work: compression level not specified, track 1 setting compression level 5.00
[17:37:32] work: only 1 chapter, disabling chapter markers
[17:37:32] job configuration:
[17:37:32]  * source
[17:37:32]    + C:\Users\Rosemary\Downloads\Source-10s.mkv
[17:37:32]    + title 1, chapter(s) 1 to 1
[17:37:32]    + container: matroska,webm
[17:37:32]    + data rate: 35869 kbps
[17:37:32]  * destination
[17:37:32]    + C:\Users\Rosemary\Downloads\Encode-QSV-H265-10bit.mkv
[17:37:32]    + container: Matroska (libavformat)
[17:37:32]  * video track
[17:37:32]    + decoder: h264_qsv 8-bit (yuv420p)
[17:37:32]    + filters
[17:37:32]      + Comb Detect (mode=3:spatial-metric=2:motion-thresh=1:spatial-thresh=1:filter-mode=2:block-thresh=40:block-width=16:block-height=16)
[17:37:32]      + Decomb (mode=39)
[17:37:32]      + Crop and Scale (width=1920:height=1036:crop-top=22:crop-bottom=22:crop-left=0:crop-right=0)
[17:37:32]        + source: 1920 * 1080, crop (22/22/0/0): 1920 * 1036, scale: 1920 * 1036
[17:37:32]      + Format (format=p010le)
[17:37:32]    + Output geometry
[17:37:32]      + storage dimensions: 1920 x 1036
[17:37:32]      + pixel aspect ratio: 1 : 1
[17:37:32]      + display dimensions: 1920 x 1036
[17:37:32]    + encoder: H.265 10-bit (Intel Media SDK)
[17:37:32]      + preset:  quality
[17:37:32]      + profile: auto
[17:37:32]      + level:   auto
[17:37:32]      + bitrate: 10000 kbps, pass: 0
[17:37:32]      + color profile: 1-1-1
[17:37:32]      + chroma location: left
[17:37:32]  * audio track 1
[17:37:32]    + decoder: ??? (DTS-HD MA) (5.1 ch) (track 1, id 0x1)
[17:37:32]      + samplerate: 48000 Hz
[17:37:32]    + mixdown: Stereo
[17:37:32]    + dither: none
[17:37:32]    + encoder: FLAC 16-bit (libavcodec)
[17:37:32]      + samplerate: 48000 Hz
[17:37:32]      + compression level: 5.00
[17:37:32] sync: expecting 247 video frames
[17:37:32] encqsvInit: using encode-only path
[17:37:32] encqsvInit: H.265/HEVC Main 10 profile @ level (null)
[17:37:32] encqsvInit: TargetUsage 4 AsyncDepth 4
[17:37:32] encqsvInit: GopRefDist 4 GopPicSize 24 NumRefFrame 3 IdrInterval 0
[17:37:32] encqsvInit: BFramesMax 3 BRefType pyramid
[17:37:32] encqsvInit: RateControlMethod VBR TargetKbps 10000 MaxKbps 10000 BufferSizeInKB 2500 InitialDelayInKB 1250 BRCParamMultiplier 1
[17:37:32] sync: first pts video is 0
[17:37:32] sync: "Chapter 1" (1) at frame 1 time 0
[17:37:32] sync: first pts audio 0x1 is 180
[17:37:33] qsv_enc_init: using 'hardware (1) via D3D11' implementation, API: 1.35
[17:37:36] reader: done. 1 scr changes
[17:37:38] work: average encoding speed for job is 47.630333 fps
[17:37:38] comb detect: heavy 0 | light 26 | uncombed 221 | total 247
[17:37:38] decomb: deinterlaced 0 | blended 26 | unfiltered 221 | total 247
[17:37:38] dca-decoder done: 966 frames, 0 decoder errors
[17:37:38] h264_qsv-decoder done: 247 frames, 0 decoder errors
[17:37:38] sync: got 247 frames, 247 expected
[17:37:38] sync: framerate min 23.976 fps, max 23.976 fps, avg 23.976 fps
[17:37:39] mux: track 0, 247 frames, 12183101 bytes, 9422.66 kbps, fifo 256
[17:37:39] mux: track 1, 108 frames, 899370 bytes, 695.59 kbps, fifo 128
[17:37:39] Finished work at: Tue Mar 14 17:37:39 2023
[17:37:39] libhb: work result = 0

 # Job Completed!
```


### What Operating System are you running?

Windows NT 10.0.19045.0, update and with latest Intel GPU driver.

### What version of HandBrake are you running?

HandBrake 1.6.1 (2023012300)

### Where did you download HandBrake from?

handbrake.fr, I believe

--- a/contrib/ffmpeg/A20-libswscale-fix-yuv420p-to-p01xle-color-conversion-bu.patch
+++ b/contrib/ffmpeg/A20-libswscale-fix-yuv420p-to-p01xle-color-conversion-bu.patch
@@ -1,0 +1,37 @@
+From dbfa7cab8c9951ec13481a9ca7f6f219dd2aba94 Mon Sep 17 00:00:00 2001
+From: Vladyslav Sosunovych <vladyslav.sosunovych@intel.com>
+Date: Wed, 26 Apr 2023 15:53:52 +0200
+Subject: [PATCH] libswscale: fix yuv420p to p01xle color conversion bug
+
+---
+ libswscale/swscale_unscaled.c | 6 +++---
+ 1 file changed, 3 insertions(+), 3 deletions(-)
+
+diff --git a/libswscale/swscale_unscaled.c b/libswscale/swscale_unscaled.c
+index 9af2e7ecc3..9f9d505d9e 100644
+--- a/libswscale/swscale_unscaled.c
++++ b/libswscale/swscale_unscaled.c
+@@ -297,7 +297,7 @@ static int planar8ToP01xleWrapper(SwsContext *c, const uint8_t *src[],
+         const uint8_t *tsrc0 = src[0];
+         for (x = c->srcW; x > 0; x--) {
+             t = *tsrc0++;
+-            output_pixel(tdstY++, t | (t << 8));
++            output_pixel(tdstY++, (t << 8));
+         }
+         src[0] += srcStride[0];
+         dstY += dstStride[0] / 2;
+@@ -308,9 +308,9 @@ static int planar8ToP01xleWrapper(SwsContext *c, const uint8_t *src[],
+             const uint8_t *tsrc2 = src[2];
+             for (x = c->srcW / 2; x > 0; x--) {
+                 t = *tsrc1++;
+-                output_pixel(tdstUV++, t | (t << 8));
++                output_pixel(tdstUV++, (t << 8));
+                 t = *tsrc2++;
+-                output_pixel(tdstUV++, t | (t << 8));
++                output_pixel(tdstUV++, (t << 8));
+             }
+             src[1] += srcStride[1];
+             src[2] += srcStride[2];
+-- 
+2.39.0
+

